extracted/7/man/man1/cc11.1
.TH CC11 1
.SH NAME
cc11 \- cross compiler for the \s-2LSI\-11\s0
.SH SYNOPSIS
.B cc11
[ option ] ... file ...
.SH DESCRIPTION
.I Cc11
is a general-purpose script that invokes the
C preprocessor, C cross-compiler,
cross-assembler, and cross-loader
to produce code for an \s-2LSI 11\s0.
\f2Cc11\f1 is easier to use than invoking the processors
individually because it passes the necessary arguments
like library directories and keeps track of intermediate
files automatically.
By default, \f2cc11\f1 leaves the result in a file named `a.out'
that is ready for downloading into the \s-2LSI 11\s0 memory.
.B N.B.:
.I cc11
produces object code in the \s-2PDP\-11 UNIX\s0 format; as expected,
it is incompatible with the \s-2VAX UNIX\s0 object code format.
.PP
By default, \f2cc11\f1 preprocesses, cross-compiles, cross-assembles,
and cross-loads a program; arguments can limit
which steps it performs and change
such things as the order in which libraries are searched.
.PP
Arguments whose names end with `.c' are taken to be
C source programs; they are compiled, and
each object program is left on the file
whose name is that of the source with `.o' substituted
for `.c'.
.PP
In the same way,
arguments whose names end with `.s' are taken to be assembly source language
and are assembled, producing a `.o' file.
.PP
Arguments whose names end with `.o' are combined with
`.o' files produced from `.c' and `.s' arguments by the loader
to produce an `a.out' file.
.PP
Arguments whose names end with `.a' are taken to be libraries
of `.o' programs, and are searched by the loader in the order
specified.
.PP
Other arguments to \f2cc11\f1 control the process as follows:
.TP 8
.B \-c
Suppress the loading phase of the compilation, and force
\f2cc11\f1 to stop after producing `.o' files.
.TP
.BI \-o " output"
Name the final output file
.IR output .
.TP 8
.B \-v
Verbose output narrating
each step of the compile. This is useful if you want
to see what's going on.
.TP
.SM
.BI \-D name=def
.br
.ns
.TP
.SM
.BI \-D \*Sname
Define the
.I name
to the preprocessor,
as if by
`#define'.
If no definition is given, the name is defined as "1".
.TP                                                             
.SM                                                             
.B \-E                                                          
Run only the macro preprocessor                                 
on the named C programs, and send the result to the   
standard output.                                            
.TP
.SM
.BI \-I \*Sdir
`#include' files
whose names do not begin with `/' are always
sought first in the directory
of the
.I file
argument,
then in directories named in
.B \-I
options,
then in directories on a standard list.
.TP                                                     
.SM
.B \-L
If loading, include the standalone startup routine in preparation
for downloading without Xinu.
The default load sequence includes Xinu startup routines.
.TP
.SM
.B \-O
Invoke an
object-code improver as part of the C cross-compilation.
.TP
.SM
.B \-P
Run only the macro preprocessor
on the named C programs, and send the result to corresponding
files suffixed `.i'.
.TP
.SM
.B \-S
Compile the named C programs into assembly language, and leave the
assembler-language output on corresponding files suffixed `.s',
without assembling or loading the result.
.TP
.SM
.BI \-U \*Sname
Remove any initial definition of
.IR name .
.PP
Other arguments
are taken
to be either loader option arguments, or C-compatible
object programs, typically produced by an earlier
.I cc11
run,
or perhaps libraries of C-compatible routines.
These programs, together with the results of any
compilations specified, are loaded (in the order
given) to produce a file that may be downloaded to the
\s-2LSI\-11\s0 for execution.
Unless changed with the \f2\-o\f1 option, \f2cc11\f1 leaves
the result of the load in file `a.out'.
.SH EXAMPLES
To compile a C program and load it with the Xinu libraries, producing
an image in `a.out' that is ready for downloading:
.sp
.ce
cc11 program.c
.sp
To compile C code in file f1.c and leave the object code in file f1.o:
.sp
.ce
cc11 \-c f1.c
.sp
To assemble code in file f2.s and leave the object code in file f2.o:
.sp
.ce
cc11 \-c f2.s
.sp
To compile C code in file f3.c and combine it with previously compiled
code in file f1.o:
.sp
.ce
cc11 f3.c f1.o
.sp
To load three previously compiled files f1.o, f2.o, and f3.o:
.sp
.ce
cc11 f1.o f2.o f3.o
.sp
To compile and load the C code in file f4.c leaving the output in `outfile'
instead of `a.out':
.sp
.ce
cc11 \-o outfile  f4.c
.sp
.ta 1.75i
.SH FILES
file.c	input C file
.br
file.s	input or output assembler language file
.br
file.o	input or output object file
.br
a.out	loaded output
.br
/tmp/ctm?	temporary compiler files
.br
Xinu-directory/lib/cpp11	preprocessor
.br
Xinu-directory/lib/c0.11	first pass
.br
Xinu-directory/lib/c1.11	second pass
.br
Xinu-directory/lib/c2.11	optional optimizer
.br
Xinu-directory/include	`#include' files
.SH DIAGNOSTICS
The diagnostics produced by C itself are intended to be
self-explanatory.
Occasional messages may be produced by the assembler
or loader.
.SH BUGS
.I Cc11
does not support arguments \f2-g\f1, \f2-w\f1, \f2-p\f1,
\f2-R\f1, \f2-C\f1, \f2-B\f1, or \f2-t\f1 that the VAX C compiler handles.
.SH SEE ALSO
as11(1), ld11(1), ranlib11(1), download(1), ar(1)


extracted/7/src/cmd/cc11/cc11.c
/* cc11.c - main */

#include <stdio.h>
#include <sys/types.h>
#include <signal.h>

/*----------------------------------------------------------------------
 * cc11 -- replacement for (and translated directly from) cc11.sh
 *----------------------------------------------------------------------
 */

char	DIR[]		= "/usr/Xinu",
	CPP[]		= "/usr/Xinu/lib/cpp11",
	C0[]		= "/usr/Xinu/lib/c0.11",
	C1[]		= "/usr/Xinu/lib/c1.11",
	C2[]		= "/usr/Xinu/lib/c2.11",
	AS[]		= "/usr/Xinu/bin/as11",
	ASDEFS[]	= "/usr/Xinu/include/sys.s",
	LD[]		= "/usr/Xinu/bin/ld11",
	CRT[]		= "/usr/Xinu/lib/xinu.o",
	*crtptr = CRT,
	LIBS[]		= "/usr/Xinu/lib/libx.a /usr/Xinu/lib/libxc.a /usr/Xinu/lib/libx.a",
	*libptr = LIBS,
	STANDLIB[]	= "/usr/Xinu/lib/standalone/libsa.a",
	STANDCRT[]	= "/usr/Xinu/lib/standalone/lowcore.o /usr/Xinu/lib/standalone/startup.o";

char *index(), *rindex();

int cflag = 1, Eflag = 0, Sflag = 0, Oflag = 0, vflag = 0, Lflag = 0;
int Aflag = 1, errors = 0;
char ldflags[256], Iflags[256], libs[512];
char cfiles[512], sfiles[512], ofiles[512];
char *ofile, *usymbol, *entry = "start", *Pflag = "", *dot, *file;
char tmp[7][20], y[160], z[160];
char pobuf[160], *cppo, sysbuf[512], *cco, *asi;
int PID = -1;


struct sig {
	int snum;
	char *sname;
	} sigs[] = {
		SIGHUP, 	"Hangup",
		SIGINT,		"Interrupt",
		SIGQUIT,	"Quit",
		SIGILL,		"Illegal Instruction",
		SIGFPE,		"Floating Point Exception",
		SIGBUS,		"Bus Error",
		SIGSEGV,	"Segmentation Violation",
		SIGSYS,		"Bad System Call",
		SIGPIPE,	"Broken Pipe",
		SIGALRM,	"Alarm Clock",
		SIGTERM,	"Terminate",
		SIGURG,		"Urgent Condition",
		SIGXCPU,	"Exceeded CPU Limit",
		SIGXFSZ,	"Exceeded File Size Limit",
		NULL,		NULL};

remove(s)
int s;
{
int i;

if (s > 0)
	signal(s,SIG_IGN);
for (i = 0; i < 7; i++)
	unlink(tmp[i]);
if (s > 0) {
	if (PID > 0)
		kill(PID,SIGTERM);
	for (i = 0; (sigs[i].snum > 0) && (sigs[i].snum < s); i++);
	fprintf(stderr,"cc11: received signal %d (%s)\n",s,sigs[i].sname);
	}
exit(s != 0);
}


main(argc,argv)
int argc;
char *argv[];
{
char **a;
int i, pid;

pid = getpid();
usymbol = "";
ofile = "a.out";
*cfiles = *sfiles = *ofiles = '\0';
*libs = *Iflags = *ldflags = '\0';
for (a = &argv[1]; *a != NULL; a++) {
	if (**a == '-') {
		switch ((*a)[1]) {
		case 'o':
			if (*++a == NULL) {
				fprintf(stderr,"-o what?\n");
				exit(1);
				}
			ofile = *a;
			break;
		case 'u':
			if (*++a == NULL) {
				fprintf(stderr,"-u what?\n");
				exit(1);
				}
			usymbol = *a;
			break;
		case 'e':
			if (*++a == NULL) {
				fprintf(stderr,"-e what?\n");
				exit(1);
				}
			entry = *a;
			break;
		case 'P':
			Pflag = "-P";
			cflag = 0;
			break;
		case 'E':
			Eflag++;
			cflag = 0;
			break;
		case 'S':
			Sflag++;
			cflag = 0;
			break;
		case 'O':
			Oflag++;
			break;
		case 'c':
			cflag = 0;
			break;
		case 'v':
			vflag++;
			break;
		case 'L':
			Lflag++;
			break;
		case 'A':
			Aflag = 0;
			break;
		case 'd':
		case 'r':
		case 's':
		case 'x':
		case 'X':
			strcat(ldflags," ");
			strcat(ldflags,*a);
			break;
		case 'i':
		case 'n':
			fprintf(stderr,"warning: can't share text on LSI-11\n");
			break;
		case 'D':
		case 'I':
		case 'U':
			strcat(Iflags," ");
			strcat(Iflags,*a);
			break;
		case 'l':
			strcat(libs," ");
			strcat(libs,*a);
			break;
		default:
			fprintf(stderr,"-%c option ignored.\n",(*a)[1]);
		}
		}
	else {
		if ( (dot = rindex(*a,'.')) == NULL) { /* object file */
			strcat(ofiles," ");
			strcat(ofiles,*a);
			continue;
			}
		switch (dot[1]) {
		case 'c':
			strcat(cfiles," ");
			strcat(cfiles,*a);
			break;
		case 's':
			strcat(sfiles," ");
			strcat(sfiles,*a);
			break;
		case 'a':
			strcat(libs," ");
			strcat(libs,*a);
			break;
		default:
			strcat(ofiles," ");
			strcat(ofiles,*a);
		}
		}
	}
strcat(Iflags," -I");
strcat(Iflags,DIR);
strcat(Iflags,"/include");
strcat(Iflags," -I");
strcat(Iflags,DIR);
strcat(Iflags,"/src/sys/h");
for (i = 0; i < 7; i++) {
	sprintf(tmp[i],"/tmp/ctm%da%d",i,pid);
	unlink(tmp[i]);
	}

for (i = 0; sigs[i].snum > 0; i++)
	signal(sigs[i].snum,remove);

for (file = cfiles+1; file != NULL && *file != '\0'; file = index(file,' ')+1) {
	dot = index(file,'.');
	*dot = '\0'; strcpy(y,file); *dot = '.';
	dot[2] = '\0'; strcpy(z,file); dot[2] = ' ';
	printf("%s:\n",z);
	if (Eflag)
		cppo = "-";
	else if (*Pflag) {
		sprintf(pobuf,"%s.i",y);
		cppo = pobuf;
		}
	else
		cppo = tmp[3];
	if (vflag)
		fprintf(stderr,"[preprocessor] %s %s %s %s\n",z,cppo,Iflags,Pflag);
	sprintf(sysbuf,"%s %s %s %s %s -DXinu",CPP,z,cppo,Iflags,Pflag);
	if (system(sysbuf)) {
		errors = 1;
		continue;
		}
	if (*Pflag)
		continue;
	if (vflag)
		fprintf(stderr,"[pass1] %s %s %s\n",cppo,tmp[0],tmp[1]);
	sprintf(sysbuf,"%s %s %s %s",C0,cppo,tmp[0],tmp[1]);
	if (system(sysbuf)) {
		errors = 1;
		continue;
		}
	if (Sflag) {
		sprintf(pobuf,"%s.s",y);
		asi = pobuf;
		if (Oflag)
			cco = tmp[4];
		else
			cco = pobuf;
		}
	else {
		if (Oflag)
			cco = tmp[4];
		else
			cco = tmp[2];
		asi = tmp[2];
		}
	if (vflag)
		fprintf(stderr,"[pass2] %s %s %s\n",tmp[0],tmp[1],cco);
	sprintf(sysbuf,"%s %s %s %s",C1,tmp[0],tmp[1],cco);
	if (system(sysbuf)) {
		errors = 1;
		continue;
		}
	if (Oflag) {
		if (vflag)
			fprintf(stderr,"[optimizer] %s %s\n",cco,asi);
		sprintf(sysbuf,"%s %s %s",C2,cco,asi);
		if (system(sysbuf)) {
			errors = 1;
			continue;
			}
		}
	if (Aflag) {
		sprintf(sysbuf,"cp %s %s;%s/bin/cvt %s > %s",
			asi,tmp[6],DIR,tmp[6],asi);
		if (vflag)
			fprintf(stderr,"[convert ps] %s\n", sysbuf);
		if (system(sysbuf)) {
			fprintf(stderr, "ps undeclared\n");
			errors = 1;
			continue;
			}
		}
	if (Sflag)
		continue;
	assemble(asi);
	}
for (file = sfiles+1; file != NULL && *file != '\0'; file = index(file,' ')+1) {
	dot = index(file,'.');
	*dot = '\0'; strcpy(y,file); *dot = '.';
	dot[2] = '\0'; strcpy(z,file); dot[2] = ' ';
	printf("%s:\n",z);
	assemble(z);
	}

if (Lflag) {
	libptr = STANDLIB;
	crtptr = STANDCRT;
}
if ( (errors == 0) && (cflag) ) {
	if (vflag)
		fprintf(stderr,"[loader] -e %s -o %s %s %s%s %s %s %s %s\n",
			entry,ofile,crtptr,*usymbol?"-u ":"",usymbol,
			ldflags,ofiles,libs,libptr);
	sprintf(sysbuf,"%s -X -e %s -o %s %s %s%s %s %s %s %s",
		LD,entry,ofile,crtptr,*usymbol?"-u ":"",usymbol,
		ldflags,ofiles,libs,libptr);
	if (system(sysbuf))
		errors = 1;
	}
remove( (errors != 0) ? -1 : 0);
}


char asbuf[160], *aso;

assemble(asi)
char *asi;
{

sprintf(asbuf,"%s.o",y);
aso = asbuf;
if (vflag)
	fprintf(stderr,"[assembler] -u -o %s %s\n",aso,asi);
sprintf(sysbuf,"%s -u -o %s %s %s %s",AS,aso,ASDEFS,asi);
if (system(sysbuf)) {
	errors = 1;
	return;
	}
if (Sflag == 0) {
	strcat(ofiles," ");
	strcat(ofiles,aso);
	}
}



system(s)
char *s;
{
	int status, pid, w;
/*	register int (*istat)(), (*qstat)(); */

	if ((pid = vfork()) == 0) {
		execl("/bin/sh", "sh", "-c", s, 0);
		_exit(127);
	}
/*	istat = signal(SIGINT, SIG_IGN); 	*** not needed for 4.2
	qstat = signal(SIGQUIT, SIG_IGN);	*/

	PID = pid;				/* for use in signal catch */
	while ((w = wait(&status)) != pid && w != -1)
		;
	PID = -1;
	if (w == -1)
		status = -1;
/*	signal(SIGINT, istat);
	signal(SIGQUIT, qstat);	*/
	return(status);
}

